include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -378,7 +378,6 @@ class EnablePreconditionable : public Preconditionable {
  * @ingroup solver
  * @ingroup LinOp
  */
-template <typename MatrixType = LinOp>
```

Comment:
it's a public interface break  and we use it for LowerTrs at least

Reply:
It breaks interface, yes. But IMO that interface was pretty useless. And I've fixed the use case of LowerTrs. So, for us the change is fine, since the `EnableSolverBase` has the concrete LinOp type.

Reply:
Does it introduce any bug? If not, we should keep the interface

Reply:
With the current interface it is not possible to get the stored system matrix from a solver passed in as a `LinOp`. You would need to try casting the LinOp to `SolverBase<T>` for different types. And for a custom solver with a custom user type, it would be completely impossible to get the operator.

Reply:
without breaking interface, you can make `SolverBase<MatrixType>` inherit from `SolverLinOpBase`
`SolverLinOpBase` provides the following interface (maybe you need `_linop` suffix)

Reply:
I can do that, but that is also not ideal. For that, I would deprecate `SolverBase<...>`, because it serves no purpose. The base logic (e.g. the workspace) is handled in `SolverBaseLinOp`, while the concrete matrix type is known in `EnableSolverBase`. But we still need to derive from `SolverBase<...>` in `EnableSolverBase` to keep the interface stability you propose, which will result in a lot of deprecation warnings.
IMO, breaking the interface here is the cleaner approach. I doubt that part of our interface has been used anyway, because of the issue I mentioned, and users should not derive directly from `SolverBase`.

Reply:
you do not need to deprecate `SolverBase<>` because it's still in use.
It's a public interface without any detail/internal namespace, so I can not say it is not used.
(unless we change the definition)
If it is the bug, we must fix it with breaking interface sadly. Other than the bug, we should keep the public interface as much as possible although it introduces much pain.
From the current changelog, we only change the behavior, not the interface (multigrid, executor) and the experimental feature (distributed)

Reply:
IMO `SolverBase<...>` needs to be deprecated, because it should be removed. We could also remove it without deprecation, but I find that can cause even more issues on the user side.
I know that it might be used, but I argued that I find that extremely unlikely.

Reply:
I've now kept the old interface and deprecated it. To reduce clutter, I disabled the corresponding warning for the header.